snewpdag/plugins/DiffPointing.py

```diff
@@ -0,0 +1,195 @@
+    logging.info('dp = {}'.format(dp))
+    return d
+
+  def weight_matrix(self, keys):
+    """
+    Calculate weight matrix.
+    Arguments:
+      keys = ordered list of keys of (det1, det2).
+    Returns matrix as np.array, columns/rows ordered as in keys.
+    """
+    dim = len(self.cache)
+    v = np.zeros([dim, dim])
+    i = 0
+    for k1 in keys:
+      d1 = self.cache[k1]
+      j = 0
+      for k2 in keys:
+        d2 = self.cache[k2]
+        if i == j:
+          v[i,j] = d2['var']
+        else:
+          if k1[0] == k2[0]:
+            v[i,j] = d1['dsig1'] * d2['dsig1']
+          elif k1[1] == k2[1]:
+            v[i,j] = d1['dsig2'] * d2['dsig2']
+          elif k1[0] == k2[1]:
+            v[i,j] = d1['dsig1'] * d2['dsig2']
+          elif k1[1] == k2[0]:
+            v[i,j] = d1['dsig2'] * d2['dsig1']
+        j += 1
+      i += 1
+    # then invert the matrix
+    logging.info('covariance matrix = {}'.format(v))
+    return np.linalg.inv(v)
+
+  def reevaluate(self, data):
+    """
+    Reevaluate direction based on available time differences
+    """
+    keys = self.cache.keys() # keep list to preserve order
+    w = self.weight_matrix(keys) # shape [nkeys,nkeys]
+
+    m = np.zeros(self.npix)
+    rs = hp.pixelfunc.pix2vec(self.nside, range(self.npix), nest=True)
+    # rs will be an np.array of x,y,z values, each triple a unit vector.
+    # however, it'll be returned in shape [3,npix]
+    d = self.d_vectors(keys, rs) # returns shape [npix,nkeys]
+    dw = d @ w # returns shape [npix,nkeys]
+    for i in range(self.npix):
+      m[i] = np.dot(dw[i], d[i])
+
+    chi2_min = m.min()
+    m -= chi2_min
+    data['map'] = m
+    data['ndof'] = 2
+    return data
+
+  def alert(self, data):
+    if 'dts' in data: # dictionary of time differences
+      for k in data['dts']:
+        # verify that the detector is in the database
+        if self.db.has(k[0]) and self.db.has(k[1]):
+          # check if it's already in the cache. If so, delete it.
+          if k in self.cache:
+            self.cache.pop(k)
+          else:
+            krev = (k[1], k[0]) # reverse order
+            if krev in self.cache:
+              self.cache.pop(krev)
+
+          nrow = self.cache_values(k[0], k[1], data['dts'][k])
+          if nrow != None:
+            self.cache[k] = nrow
+
+    if len(self.cache) >= self.min_dts:
+      return self.reevaluate(data)
+    else:
+      return False
+
+  def revoke(self, data):
+    if 'dts' in data:
+      for k in data['dts']:
+        # verify that the detector is in the database
+        if self.db.has(k[0]) and self.db.has(k[1]):
+          # check if it's already in the cache. If so, delete it.
+          if k in self.cache:
+            self.cache.pop(k)
+          else:
+            krev = (k[1], k[0]) # reverse order
+            if krev in self.cache:
+              self.cache.pop(krev)
+
+    if len(self.cache) >= self.min_dts:
+      return self.reevaluate()
```

Comment:
I suspect we are missing a data in brackets here

Reply:
I think you're right